### PR TITLE
feat(native): Add option to limit crash reports per group

### PR DESCRIPTION
**Pull request by jan-auer**
Adds a feature to limit crash reports per issue group.

This extends the existing `sentry:store_crash_reports` option to introduce a per-issue limit. The old option only toggled storage of crash reports on or off, the new one allows to set a limit. The new setting has three states:
 - `0` disables storage of crash reports
 - `-1` enables unbounded storage of crash reports
 - any positive number sets the limit.
- (for backwards compatibility, the old `True` value is now shimmed to `-1`)

To avoid database queries in every store request, we cache the number of crash reports per issue group once we've reached the limit. This also works with group merging and unmerging. Note that we do not prune the cache keys, since merging and unmerging does not invalidate the cache invariant.

This is inherently racy between our store workers. Therefore, the count is just a best-effort to limit the number overall, but might lead to slightly more crash reports than what's configured. The obvious exception is `0`, where crash reports are not stored at all.

## New Settings UI

image

The old setting is applied automatically.

## Migration

To run the count query efficiently, this requires `group_id` to be set on event attachments. Unfortunately, this was not done in the past, which is why there is a migration in https://github.com/getsentry/sentry/pull/15801.

### CODE REVIEW

**Comment by jan-auer** [comment]
The migration is still missing. I'll add it momentarily.

File: src/sentry/tasks/store.py [hunk 1]
```diff
@@ -417,16 +425,104 @@ def save_attachment(event, attachment):
...
+        cached_reports = 0
+    stored_reports = cached_reports
+
+    for attachment in attachments:
+        # If the attachment is a crash report (e.g. minidump), we need to honor
+        # the store_crash_reports setting. Otherwise, we assume that the client
+        # has already verified PII and just store the attachment.
+        if attachment.type in CRASH_REPORT_TYPES:
+            if crashreports_exceeded(stored_reports, max_crashreports):
+                continue
```
Comment (untitaker) [inline src/sentry/tasks/store.py]: can this be a `break`?

File: src/sentry/tasks/store.py [hunk 2]
```diff
@@ -417,16 +425,104 @@ def save_attachment(event, attachment):
             if not organization.get_option("sentry:store_crash_reports"):
                 return
 
-    file = File.objects.create(
```
Comment (untitaker) [inline src/sentry/tasks/store.py]: Remove the rest of this func

**Review by untitaker — APPROVED** [review]

File: src/sentry/tasks/store.py (hunk 2 above)
Comment (jan-auer) [reply src/sentry/tasks/store.py]: Done.

File: src/sentry/tasks/store.py (hunk 1 above)
Comment (jan-auer) [reply src/sentry/tasks/store.py]: We still need to continue with the next attachments, since there might be non-crash-report attachments in the list.

**Comment by jan-auer** [comment]
Adding @JTCunning for the Event attachment query. Note that we cache the limit once a group has exceeded it. So we only run this for the first events of each group.

https://github.com/getsentry/sentry/pull/15798/files#diff-56a31c3653d159cb70168fe95ec5958fR436-R438

**Comment by jan-auer** [comment]
Offline review: Change to django cache.

**Review by tonyo — CHANGES_REQUESTED** [review]
As discussed internally, let's use memcached here.

**Comment by jan-auer** [comment]
@tonyo updated to Django cache.

**Review by tonyo — APPROVED** [review]

**Pull request merged**